### PR TITLE
[RNMobile] Embed Block caption field

### PR DESCRIPTION
**Pull request by jd-alexander**
## Description

This PR implements the Embed Block caption functionality. 

Gutenberg Mobile PR: https://github.com/wordpress-mobile/gutenberg-mobile/pull/3552

**⚠️ IMPORTANT NOTE:** **[✅  Done in [this commit](https://github.com/WordPress/gutenberg/pull/32226/commits/9f14b7e2c7744cd5a8690b35c47e379b48c44573)]** The following code blocks were added only for testing purposes, so once the PR is approved they should be removed/reverted before merging.

- https://github.com/WordPress/gutenberg/blob/93def617478dfa20a9a2a0b93ae629916a4b3d67/packages/block-library/src/embed/edit.js#L87-L92

## How has this been tested?
**Note:** Currently, the `EmbedPreview` is not available so a `Text` component simulates it's position within the container. The Embed Preview will be implemented here wordpress-mobile/gutenberg-mobile#3276

### Renders properly

1. Add an Embed block. 
2. Notice that the `Embed Placeholder` is disabled so the `EmbedPreview` can be rendered without setting a link. ( Link setting was implemented in https://github.com/wordpress-mobile/gutenberg-mobile/pull/3436
3. Notice that the caption field is rendered and selected.
4. Verify that unselecting the Embed block causes the empty caption to disappear. 
5. Verify that unselecting the Embed block that has a caption, causes the caption's `RichText` component to lose focus. 
6. Ensure that pressing enter at the end of a caption entry causes a new paragraph to be created.

### HTML output is valid
1. Add an Embed block. 
2. Notice that the `Embed Placeholder` is disabled so the `EmbedPreview` can be rendered without setting a link. ( Link setting was implemented in https://github.com/wordpress-mobile/gutenberg-mobile/pull/3436
3. Type some text in the caption field.
4. Put the caret in the middle of the text and press Enter to add a new line.
5. Check on HTML mode that the resulting HTML code is correct, the caption text should have `<br>` on line-breaks.

### iOS Light Mode

Empty caption | Caption with value 
--------|-------
 <kbd></kbd>    |   <kbd></kbd>

### iOS Dark Mode

Empty caption | Caption with value 
--------|-------
 <kbd></kbd>    |   <kbd></kbd>

### Android Light Mode

Empty caption | Caption with value 
--------|-------
 <kbd></kbd>    |   <kbd></kbd>

### Android Dark Mode

Empty caption | Caption with value 
--------|-------
 <kbd></kbd>    |   <kbd></kbd>

## Types of changes

- The `EmbedPreview` component was recreated here even though there's a WIP of it in https://github.com/wordpress-mobile/gutenberg-mobile/issues/3275 this was done for demonstration purposes. 

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/block-library/src/embed/edit.js [hunk 1]
```diff
@@ -193,7 +194,9 @@ const EmbedEdit = ( props ) => {
 		native: title,
 	} );
 	// No preview, or we can't embed the current URL, or we've clicked the edit button.
-	const showEmbedPlaceholder = ! preview || cannotEmbed || isEditingURL;
+	//const showEmbedPlaceholder = ! preview || cannotEmbed || isEditingURL;
+	const showEmbedPlaceholder = false;
```
Comment (fluiddot) [inline packages/block-library/src/embed/edit.js]: We have to hide the embed placeholder to test the PR but I'd revert this change once it's approved.

File: packages/block-library/src/embed/embed-preview.native.js [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+			onPress={ onEmbedPreviewPress }
+			disabled={ ! isSelected }
+		>
+			<View>
+				<View style={ styles[ 'embed-preview__placeholder' ] }>
+					<Text style={ styles[ 'embed-preview__placeholder-text' ] }>
+						Embed Preview will be directly above the Block Caption
+						component when it is implemented.
+					</Text>
+				</View>
```
Comment (fluiddot) [inline packages/block-library/src/embed/embed-preview.native.js]: This placeholder will be removed once we include the preview content which will be addressed in this issue:
https://github.com/wordpress-mobile/gutenberg-mobile/issues/3275.

File: packages/block-library/src/embed/edit.js [hunk 3]
```diff
@@ -179,6 +180,15 @@ const EmbedEdit = ( props ) => {
...
 
+	// NOTE: This effect is only required for testing the caption element in the native version,
+	// once the PR is approved this code block should be removed.
+	// Reference: https://github.com/WordPress/gutenberg/pull/32226
+	useEffect( () => {
+		if ( ! url ) {
+			setAttributes( { url: 'http://wordpress.com' } );
+		}
+	}, [] );
+
```
Comment (fluiddot) [inline packages/block-library/src/embed/edit.js]: If the block doesn't have an URL defined, the HTML related to the caption attribute is not displayed in the HTML mode. Since we don't have yet the ability to set the URL (will be added in https://github.com/wordpress-mobile/gutenberg-mobile/pull/3436), I added this temporary hack to let us test the PR, once the changes are approved we'll remove this code block.

**Comment by fluiddot** [comment]
@hypest The PR is ready to review, most of the work was already done so I only applied some improvements and conducted testing. So far everything looks fine but I'd like to note that there're a couple of code blocks that have been included only for testing purposes (described in the PR description), once the PR is approved we should remove them before merging.

**Comment by hypest** [comment]
**Renders properly**
✅ Add an Embed block.
✅ Notice that the Embed Placeholder is disabled so the EmbedPreview can be rendered without setting a link. ( Link setting was implemented in Embed Block Bottom Sheet to set a link wordpress-mobile/gutenberg-mobile#3436

⚠️ Notice that the caption field is rendered and selected. ➡️ _On the Android demo app, I don't see the caption being selected. The block is indeed selected, but the caption is not (the caret is not in the caption, nor the virtual keyboard is popped up). Should it be?_

✅ Verify that unselecting the Embed block causes the empty caption to disappear.
✅ Verify that unselecting the Embed block that has a caption, causes the caption's RichText component to lose focus.
✅ Ensure that pressing enter at the end of a caption entry causes a new paragraph to be created.

---

**HTML output is valid**

✅ Add an Embed block.
✅ Notice that the Embed Placeholder is disabled so the EmbedPreview can be rendered without setting a link. ( Link setting was implemented in Embed Block Bottom Sheet to set a link wordpress-mobile/gutenberg-mobile#3436
✅ Type some text in the caption field.
✅ Put the caret in the middle of the text and press Enter to add a new line.
✅ Check on HTML mode that the resulting HTML code is correct, the caption text should have <br> on line-breaks.

File: packages/block-library/src/embed/edit.js (hunk 3 above)
Comment (hypest) [reply packages/block-library/src/embed/edit.js]: The Gutenberg side PR of the url-set feature is merged at this point, so wdyt about removing the test code and rely on setting the url as normal?

File: packages/block-library/src/embed/edit.js [hunk 4]
```diff
@@ -279,6 +294,7 @@ const EmbedEdit = ( props ) => {
 					icon={ icon }
 					label={ label }
 					insertBlocksAfter={ insertBlocksAfter }
+					clientId={ clientId }
```
Comment (hypest) [inline packages/block-library/src/embed/edit.js:284]: Interesting that we need to pass the block id so the [caption component can determine](https://github.com/WordPress/gutenberg/blob/80465ea0ef7f804e515bb1b0137a31890bd7baaf/packages/block-editor/src/components/block-caption/index.native.js#L50-L56) if it should display itself. Nothing to touch in this PR but, I intuitively feel that displaying the caption or not should be the caption's parent responsibility, and in this case only governed by the `isSelected`. Anyway, just a comment for the time being since it caught my attention.

File: packages/block-library/src/embed/edit.js (hunk 3 above)
Comment (fluiddot) [reply packages/block-library/src/embed/edit.js]: Right, but this code is preventing the embed placeholder to be rendered so we still need it for testing.
https://github.com/WordPress/gutenberg/blob/223c4407ab9c4854648ce10939698a5765d3fa24/packages/block-library/src/embed/edit.js#L225-L230

If I remove this part too then the embed preview component (which contains the caption field) won't be rendered 🙃 , at this point, we could explore a workaround for rendering a fake preview, wdyt?

**Comment by fluiddot** [comment]
> ⚠️ Notice that the caption field is rendered and selected. ➡️ On the Android demo app, I don't see the caption being selected. The block is indeed selected, but the caption is not (the caret is not in the caption, nor the virtual keyboard is popped up). Should it be?

Good catch! I'll review this part on Android, we're using the `BlockCaption` component underneath for the text field rendering, so I'll investigate this deeper just in case I overlooked any parameter 🤔 .

File: packages/block-library/src/embed/edit.js (hunk 4 above)
Comment (fluiddot) [reply packages/block-library/src/embed/edit.js:284]: Yeah, I'm surprised too that the `BlockCaption` component requires the `clientId` to determine if the block is selected, I agree that looks more intuitive if we had a `isSelected` prop here instead `clientId`.

File: packages/block-library/src/embed/edit.js (hunk 3 above)
Comment (fluiddot) [reply packages/block-library/src/embed/edit.js]: In [this commit](https://github.com/WordPress/gutenberg/pull/32226/commits/93def617478dfa20a9a2a0b93ae629916a4b3d67) I added a workaround for this case, instead of trying the get the embed preview, it's returning the URL as the preview. This way we could have both embed sheet and preview components working together, but I'd like to note that it's just a temporary solution and only for testing, there're other features like replacing the URL that won't work.

**Comment by hypest** [comment]
> so I'll investigate this deeper just in case I overlooked any parameter 🤔

tbh, I don't feel that's enough of a blocker for this PR, if the caption doesn't get selected. I might even argue that it shouldn't be selected upon adding a new embed block since the caption is not the primary interaction entry point for this particular block (the URL is).

File: packages/block-library/src/embed/edit.js (hunk 4 above)
Comment (hypest) [reply packages/block-library/src/embed/edit.js:284]: > ... if we had a isSelected prop here instead clientId.

Hehe, we do have iSelected passed, 4 lines above this one :giggle:

And just to stay clear, let's leave it as is for this PR.

File: packages/block-library/src/embed/edit.js (hunk 3 above)
Comment (hypest) [reply packages/block-library/src/embed/edit.js]: Ah, I like that "cleaner" intermediate solution with the URL. Let's keep it until the follow up PRs that will wrangle a better preview fallback, or if you want, remove it alltogether before merging this PR, up to you.

**Review by hypest — APPROVED** [review]
LGTM! Awesome work here @jd-alexander and @fluiddot !

Feel free to remove the testing code if you want, before merging.
